src/render.c

```diff
@@ -14,38 +14,110 @@
+ */
+int cmark_node_replace_with_children(cmark_node *oldnode, cmark_node *newnode) {
+  cmark_node_replace(oldnode, newnode);
+
+  cmark_node *child;
+  cmark_node *next;
+
+  while ((child = cmark_node_first_child(oldnode)) != NULL) {
+    next = cmark_node_next(child);
+    if (!cmark_node_append_child(newnode, child)) {
+      return 0;
+    }
+    child = next;
+  }
+
+  return 1;
+}
+
+/**
+ * Extracts id attribute from header and attaches it to the node as custom
+ * metadata.
+ */
+int replace_header_attributes(cmark_node *node, char *input_file_path,
+                              AnchorLocationArray anchor_locations) {
+  if (cmark_node_get_type(node) != CMARK_NODE_HEADER)
+    return 0;
+  cmark_node *last_child = cmark_node_last_child(node);
+
+  if (cmark_node_get_type(last_child) != CMARK_NODE_TEXT)
+    return 0;
+
+  const char *lit = cmark_node_get_literal(last_child);
   char *id = malloc(strlen(lit) + 1);
   int pos = parse_heading_bracketed_span_id(lit, id);
   if (-1 == pos) {
     free(id);
     return 0;
   }
-  char *new_lit = malloc(strlen(lit) + 1);
-  int i;
-  for (i = 0; i < pos; i++) {
-    new_lit[i] = lit[i];
+
+  HeadingInfo *info = malloc(sizeof *info);
+  info->anchor = malloc(strlen(id) + 1);
+  strcpy(info->anchor, id);
+  cmark_node_set_user_data(node, info);
+
+  // TODO: populate anchor_locations here
+
+  if (pos == 0) {
+    cmark_node_unlink(last_child);
+  } else {
+    char *new_lit = malloc(strlen(lit) + 1);
+    int i;
+    for (i = 0; i < pos; i++) {
+      new_lit[i] = lit[i];
+    }
+    new_lit[i] = '\0';
+    cmark_node *new_node = cmark_node_new(CMARK_NODE_TEXT);
+    cmark_node_set_literal(new_node, new_lit);
+    cmark_node_replace(last_child, new_node);
   }
-  new_lit[i] = '\0';
-  char *first_span = malloc(12 + strlen(lit) + 1);
-  strcpy(first_span, "<span id='");
-  strcat(first_span, id);
-  strcat(first_span, "'>");
-  cmark_node *new_node = cmark_node_new(CMARK_NODE_HTML_INLINE);
-  cmark_node_set_literal(new_node, first_span);
-  cmark_node_insert_before(node, new_node);
-  new_node = cmark_node_new(CMARK_NODE_HTML_INLINE);
-  cmark_node_set_literal(new_node, "</span>");
-  cmark_node_insert_after(node, new_node);
-  new_node = cmark_node_new(CMARK_NODE_TEXT);
-  cmark_node_set_literal(new_node, new_lit);
-  cmark_node_insert_after(node, new_node);
-  cmark_node_replace(node, new_node);
-  cmark_node_free(node);
+
+  cmark_node_free(last_child);
   free(id);
-  free(new_lit);
-  free(first_span);
+  return 1;
+}
+
+/**
+ * Converts headers with custom metadata into HTML heading node.
+ */
+int replace_headers_with_attributes_for_html(cmark_node *node) {
+  if (cmark_node_get_type(node) != CMARK_NODE_HEADER)
+    return 0;
+
+  HeadingInfo *info = cmark_node_get_user_data(node);
+
+  if (!info)
+    return 0;
+
+  char level[2];
+  snprintf(level, sizeof(level), "%d", cmark_node_get_heading_level(node));
+  char *on_enter = malloc(2 + 1 + 5 + strlen(info->anchor) + 1);
```

Comment:
```suggestion
  char *on_enter = malloc(2 + 1 + 5 + strlen(info->anchor) + 2 + 1);
```

Reply:
Thanks. That fixes the issues detected by `meson test -C build --wrap=valgrind`. (Except for the one that is also present on main.